dcplib/etl/__init__.py

```diff
@@ -0,0 +1,128 @@
+            sys.stdout.write(".")
+            sys.stdout.flush()
+        return f, bundle_uuid, bundle_version
+
+    def get_files_to_fetch_for_bundle(self, bundle_uuid, bundle_version, content_type_patterns):
+        logger.debug("[%s] Fetching manifest for bundle %s", threading.current_thread().getName(), bundle_uuid)
+        res = http.get(f"{self.dss_host}/bundles/{bundle_uuid}", params={"replica": "aws"})
+        res.raise_for_status()
+        logger.debug("Scanning bundle %s", bundle_uuid)
+        files_to_fetch = []
+        for f in res.json()["bundle"]["files"]:
+            if any(fnmatchcase(f["content-type"], p) for p in content_type_patterns):
+                os.makedirs(f"{self.sd}/bundles/{bundle_uuid}.{bundle_version}", exist_ok=True)
+                try:
+                    with open(f"{self.sd}/files/{f['uuid']}", "rb") as fh:
+                        file_csum = hashlib.sha256(fh.read()).hexdigest()
+                        if file_csum == f["sha256"]:
+                            logger.debug("Linking fetched file %s/%s", bundle_uuid, f["name"])
+                            if not os.path.exists(f"{self.sd}/bundles/{bundle_uuid}.{bundle_version}/{f['name']}"):
+                                os.symlink(f"../../files/{f['uuid']}",
+                                           f"{self.sd}/bundles/{bundle_uuid}.{bundle_version}/{f['name']}")
+                            continue
+                except (FileNotFoundError, ):
+                    pass
+                files_to_fetch.append(f)
+            else:
+                logger.debug("Skipping non-metadata file %s/%s", bundle_uuid, f["name"])
+        return bundle_uuid, bundle_version, files_to_fetch
+
+    def enqueue_bundle_manifests(self, percent_complete, bundles_complete, total_bundles, f2f_futures, ff_futures,
+                                 executor):
+        logger.info("[%d%%] [%d/%d] Processing bundle batch",
+                    percent_complete, bundles_complete, total_bundles)
+        for future in concurrent.futures.as_completed(f2f_futures):
+            bundle_uuid, bundle_version, files_to_fetch = future.result()
+            if files_to_fetch:
+                logger.info("[%d%%] Fetching %d files for bundle %s @ %s",
+                            percent_complete, len(files_to_fetch), bundle_uuid, bundle_version)
+                for f in files_to_fetch:
+                    ff_futures.append(executor.submit(self.get_file, f, bundle_uuid, bundle_version))
+                    self.b2f[bundle_uuid + "." + bundle_version].add(f["name"])
+            else:
+                self.staged_bundles.append([bundle_uuid, bundle_version])
+        return len(f2f_futures)
+
+    def extract(self, query: dict = None, transformer: callable = None, loader: callable = None,
+                content_type_patterns=None, filename_patterns=None, max_workers=512):
+        if query is None:
+            query = self.default_bundle_query
```

Comment:
Maybe check for valid JSON on this for minimal validation

Reply:
Can you elaborate on the type of check that you envision here? The type hint already specifies JSON, and I'd rather let the client or API produce an error in `post_search()` than duplicate that logic here.

Reply:
Ah the type hint :+1: . That all sounds good